### ISSUE TITLE
Make benchmarks more realistic

### ISSUE THREAD

**Issue by lfairy** [lfairy/maud#49]
The [existing benchmarks](https://github.com/lfairy/maud/tree/master/benchmarks/benches) are okay, but a bit simplistic. I'd love to have code that reflects the kind of use cases we see in the real world.

**Comment by lfairy** [comment]
@TheNeikos would you be interested in this issue? I know you're working on a website that uses Maud; that project could be a good source to draw from.

**Comment by TheNeikos** [comment]
Hm, right now I'm doing something similar to Rails (cause I like their way of organizing things):

- controller calls view with data
- view computes it with `html!` which usually includes several other components that call `html!`
- it is all passed to a layout that then outputs the whole thing

I wouldn't mind generalizing it a bit and making a benchmark for it, although I am not sure what one would benchmark in that case. Nested `html!`? (Because benchmarking my way of doing it would be out of scope of Maud think)

**Comment by lfairy** [comment]
It's great that you're interested in this sort of thing!

My goals with these benchmarks are twofold:

1.  To give me something more realistic to benchmark against, so I don't optimize myself into a hole. We had that weird issue with inlining before, and I suspect that a larger benchmark would have been more helpful for that.

2.  To compare approaches with other template engines.

    Horrorshow, in particular, still uses a method based on writers/closures, rather than returning a `String` as Maud does today. We only see a difference in the two approaches when we start including templates in other templates (I think that's what you mean by nested `html!`?). One of the reasons I switched to `String`s was to make use cases like yours more convenient, so I'd like to see how it compares performance-wise.

In other words, benchmarking your way of doing it is *not* out of scope. I think copying the code from a reasonably-complex view, and calling it would some hard-coded data, would make a decent benchmark.

**Comment by TheNeikos** [comment]
Alright, I will do that then. :+1:

**Comment by TheNeikos** [comment]
Hm... Sorry that I didn't come back to you yet. I forgot to put it on my to-do list and it slipped my mind. It is there now though, so you should get the start of a PR tonight (GMT+2).

**Comment by lfairy** [comment]
I don't think there's anything else to do for this issue.
